src/crypto/tls/handshake_client_tls13.go

```diff
@@ -58,6 +62,14 @@ func (hs *clientHandshakeStateTLS13) handshake() error {
 	hs.transcript = hs.suite.hash.New()
 	hs.transcript.Write(hs.hello.marshal())
 
+	// When offering ECH, it is not known whether ECH was accepted until the
+	// HelloRetryRequest is processed. In particular, we do not know at this
+	// point if the server used the ClientHelloOuter or the ClientHelloInner.
+	if c.ech.offered {
+		hs.transcriptInner = hs.suite.hash.New()
+		hs.transcriptInner.Write(hs.helloInner.marshal())
+	}
+
```

Comment:
This code is required even after https://github.com/tlswg/draft-ietf-tls-esni/issues/274 lands: because the explicit signal only appears in the SH, the client does not know whether ECH was accepted until after processing the HRR.

Reply:
Oof, that HRR situation is an annoying edge case requiring extra work/memory :(

I have not thought much about it, is there no way to avoid it at the spec level? (thinking aloud)

Reply:
I don't think so. One idea is to add an acceptance indication to the HRR much like we do to the SH; but this would make the HRR path "stick out".